### PR TITLE
OCPBUGS-86669: fix(cli): make deleteCLISecrets non-fatal during cluster destroy

### PR DESCRIPTION
**Pull request by rutvik23**
## What this PR does / why we need it:
- Make `deleteCLISecrets` non-fatal in the `DestroyCluster` CLI flow to prevent the
    `openshift.io/destroy-cluster` finalizer from getting stuck when the service account
    lacks `deletecollection` RBAC permission
- Register `hyperapi.Scheme` in the fake test client to enable unit tests with HyperShift types
- Add test coverage for the destroy finalizer lifecycle with a non-nil HostedCluster

## Problem

- When running `hcp destroy cluster`, the `openshift.io/destroy-cluster` finalizer is added early in the flow and removed at the very end. The `deleteCLISecrets` function sits between `destroyPlatformSpecifics` and `removeFinalizer`:
```
  setFinalizer()              ← finalizer added
  destroyPlatformSpecifics()  ← succeeds (AWS cleanup done)
  deleteCLISecrets()          ← FAILS (RBAC: deletecollection forbidden)
  removeFinalizer()           ← NEVER REACHED
```

 - `deleteCLISecrets` uses `DeleteAllOf`, which requires the `deletecollection` RBAC verb — a more elevated permission than standard CRUD. When the calling service account lacks this permission, the function returns an error, the CLI exits, and the finalizer is never removed. The HostedCluster remains stuck with `deletionTimestamp` set indefinitely.

## Special notes for your reviewer:
 **Why non-fatal instead of removing the finalizer entirely?**
  - The `openshift.io/destroy-cluster` finalizer keeps the HostedCluster alive while the CLI
  performs platform infrastructure teardown (VPCs, IAM, DNS, etc.). If `destroyPlatformSpecifics`
  fails, the HC staying around is valuable; it serves as the source of truth for InfraID,
  Region, and BaseDomain, allowing the user to simply re-run `hypershift destroy` without
  remembering those values. Removing the finalizer entirely would lose this recovery anchor.

  **Why is `deleteCLISecrets` safe to skip?**
 - The operator's `reconcileCLISecrets` (hostedcluster_controller.go) sets ownerRefs on
  CLI-created secrets pointing to the HostedCluster. When the finalizer is removed and the HC
  is fully deleted, Kubernetes garbage collection handles secret cleanup automatically. The
  CLI's explicit `DeleteAllOf` is redundant.

  **Why change the fake client scheme?**
 - `GetClient()` with `FAKE_CLIENT=true` previously returned `fake.NewFakeClient()` using only
  the default Kubernetes scheme. This prevented any test from working with HyperShift types
  (HostedCluster, NodePool, etc.), the existing test worked around this by passing `nil` for
  the HostedCluster. Registering `hyperapi.Scheme` is a superset (includes all standard k8s
  types plus HyperShift types), so existing tests are unaffected.

## Which issue(s) this PR fixes:
Fixes: [OCPBUGS-86669](https://issues.redhat.com/browse/OCPBUGS-86669)

## Checklist:
- [x] Subject and description added to both, commit and PR.
- [x] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [x] This change includes unit tests.
- [x] Existing test passes: `When HostedCluster is nil and platform specifics provided`
- [x] New test passes: `When destroying a hosted cluster with platform specifics, it should 
       set the destroy finalizer and complete successfully`
- [x] `make verify` passes
- [x] `make test` passes

## Bug Fixes
* Cluster destruction now continues even if command secret cleanup fails, logging the issue and proceeding with the rest of the removal process.
* Destroy workflow finalizer behavior is validated more reliably before and after the operation.

## Tests
* Expanded destruction test coverage with a more realistic hosted cluster scenario and stronger assertions around destroy finalizer state.
* Added coverage to ensure destruction succeeds even when secret deletion is blocked (forbidden error).
* Improved test-mode fake client setup to use the correct API scheme.

### CODE REVIEW

**Review by jparrill — COMMENTED** [review]
Dropped some comments. Thanks!

Overall this is a solid fix — the non-fatal approach is architecturally sound and the finalizer contract is preserved. The main ask is adding a test that actually exercises the error path (see inline comment on the test file). The rest are non-blocking nits.

One thing worth tracking as a follow-up: not all CLI-created secrets carry the `safe-to-delete-with-cluster` label, so `reconcileCLISecrets` never sets ownerRefs on them (AWS SA token issuer, proxy SSH key, Azure creds, PowerVS secrets). Namespace deletion covers this, but adding the label to those secrets would close the gap properly.

File: cmd/cluster/core/destroy.go [hunk 1]
```diff
@@ -163,9 +163,11 @@ func DestroyCluster(ctx context.Context, hostedCluster *hyperv1.HostedCluster, o
 		return err
 	}
 
-	// clean up CLI generated secrets
+	// Non-fatal: secrets with ownerRef on the HostedCluster will be
+	// garbage collected when the HostedCluster is deleted.
```
Comment (jparrill) [inline cmd/cluster/core/destroy.go]: **Nit:** The comment says "secrets with ownerRef on the HostedCluster will be garbage collected," but that's only true for secrets that carry the `safe-to-delete-with-cluster` label — `reconcileCLISecrets` only sets ownerRefs on those. Platform-specific secrets (AWS SA token issuer, proxy SSH key, Azure cloud creds, PowerVS secrets) are created with `auto-created-for-infra` only, so they won't get ownerRefs.

They're still cleaned up when the operator deletes the control-plane namespace, so the non-fatal approach is safe — but the comment should reflect that. Something like:
```
// Non-fatal: qualifying secrets are garbage-collected via ownerRef when the
// HostedCluster is deleted; remaining secrets are removed when the operator
// deletes the control-plane namespace.
```

File: cmd/cluster/core/destroy.go [hunk 2]
```diff
@@ -163,9 +163,11 @@ func DestroyCluster(ctx context.Context, hostedCluster *hyperv1.HostedCluster, o
...
 	if err = deleteCLISecrets(ctx, o, c); err != nil {
-		return err
```
Comment (jparrill) [inline cmd/cluster/core/destroy.go:187]: **Nit:** There's no right or wrong answer here — `Log.Error` is semantically valid because something did fail. That said, the repo's established pattern for non-fatal errors that are expected and self-heal uses `Log.Info` with `"error", err.Error()` (13+ instances: `create.go:802`, `cmd/infra/aws/destroy.go:188`, `cmd/oadp/backup.go:122,147,151`, etc.). Just mentioning for consistency, not blocking.

File: cmd/cluster/core/destroy_test.go [hunk 3]
```diff
@@ -66,6 +71,52 @@ func TestDestroyCluster(t *testing.T) {
 		g.Expect(err).ToNot(HaveOccurred())
 		g.Expect(platformSpecificsCalled).To(BeTrue())
 	})
+
```
Comment (jparrill) [inline cmd/cluster/core/destroy_test.go:87]: **Blocking:** This test validates `setFinalizer` and the happy-path flow, which is useful — but it doesn't exercise the actual behavioral change in this PR. `deleteCLISecrets` succeeds here because the fake client has no secrets, so both the old `return err` and the new log-and-continue produce the same result.

Could you add a test where `deleteCLISecrets` actually fails? For example:
```go
t.Run("When deleteCLISecrets fails it should log and continue", func(t *testing.T) {
    g := NewGomegaWithT(t)
    t.Setenv("FAKE_CLIENT", "true")

    // Inject a Forbidden error on DeleteAllOf for Secrets
    // using fake.NewClientBuilder().WithInterceptorFuncs(...)
    // then verify DestroyCluster still returns nil
})
```
Without this, the bug fix (OCPBUGS-86669) has zero test coverage.

Comment (rutvik23) [reply cmd/cluster/core/destroy_test.go:87]: Done. Added a test that injects a Forbidden error on DeleteAllOf for Secrets via fake.NewClientBuilder().WithInterceptorFuncs(...) and verifies DestroyCluster still returns nil. Introduced a package-level var newClient in destroy.go (test seam only) so the test can override client creation. Also addressed the comment wording and switched Log.Error to Log.Info per your other suggestions.

**Comment by rutvik23** [comment]
> One thing worth tracking as a follow-up: not all CLI-created secrets carry the `safe-to-delete-with-cluster` label, so `reconcileCLISecrets` never sets ownerRefs on them (AWS SA token issuer, proxy SSH key, Azure creds, PowerVS secrets). Namespace deletion covers this, but adding the label to those secrets would close the gap properly.

Created a new [tracker](https://redhat.atlassian.net/browse/OCPBUGS-96904) to improve this further.

**Comment by rutvik23** [comment]
/verified by @rutvik23 
Manually verified, test report is attached  on the [jira](https://redhat.atlassian.net/browse/OCPBUGS-86669?focusedCommentId=17666920)

File: cmd/cluster/core/destroy.go [hunk 4]
```diff
@@ -29,6 +29,9 @@ const (
 	destroyFinalizer = "openshift.io/destroy-cluster"
 )
 
+// overridden in tests to inject a client with interceptor-based error injection
+var newClient = util.GetClientWithKubeconfig
```
Comment (csrwng) [inline cmd/cluster/core/destroy.go]: This is not needed since you use an env var in the test.

File: cmd/cluster/core/destroy.go [hunk 5]
```diff
@@ -107,7 +110,7 @@ func GetCluster(ctx context.Context, o *DestroyOptions) (*hyperv1.HostedCluster,
 func DestroyCluster(ctx context.Context, hostedCluster *hyperv1.HostedCluster, o *DestroyOptions, destroyPlatformSpecifics DestroyPlatformSpecifics) error {
 	hostedClusterExists := hostedCluster != nil
 	shouldDestroyPlatformSpecifics := destroyPlatformSpecifics != nil
-	c, err := util.GetClientWithKubeconfig(o.Kubeconfig)
+	c, err := newClient(o.Kubeconfig)
```
Comment (csrwng) [inline cmd/cluster/core/destroy.go]: my preference would be to declare an internal destroyCluster function that takes a client, and then you just pass the result of util.GetClientWithKubeconfig from the external DestroyCluster function. In your unit test  you can just call the internal function with the fake client. No need for an environment variable.

File: cmd/util/client.go [hunk 6]
```diff
@@ -59,7 +59,7 @@ func GetClient() (crclient.Client, error) {
 // the default kubeconfig resolution.
 func GetClientWithKubeconfig(kubeconfigPath string) (crclient.Client, error) {
 	if os.Getenv("FAKE_CLIENT") == "true" {
-		return fake.NewFakeClient(), nil
+		return fake.NewClientBuilder().WithScheme(hyperapi.Scheme).Build(), nil
```
Comment (csrwng) [inline cmd/util/client.go]: see my note above. If you have an internal destroyCluster function, you can construct the fake client in the unit test function and simply pass it in. There's no need to change this code.

**Review by csrwng — COMMENTED** [review]
some comments, otherwise lgtm

**Comment by rutvik23** [comment]
### Test Coverage

  Verified non-fatal `deleteCLISecrets` behavior with unit tests:

  ```
  go test -c -o /tmp/pkg_test ./cmd/cluster/core/ && \
    /tmp/pkg_test -test.v -test.count=1 -test.run "^TestDestroyCluster$"
  ```

  <details>
  <summary>Test output (all 4 subtests pass)</summary>

  ```
  --- PASS: TestDestroyCluster (0.05s)
      --- PASS: TestDestroyCluster/When_HostedCluster_is_nil_and_platform_specifics_provided_it_should_call_destroyPlatformSpecifics (0.05s)
      --- PASS: TestDestroyCluster/When_kubeconfig_is_set_it_should_use_it_for_the_client (0.00s)
      --- PASS: TestDestroyCluster/When_destroying_a_hosted_cluster_with_platform_specifics,_it_should_set_the_destroy_finalizer_and_complete_successfully
  (0.00s)
      === RUN   TestDestroyCluster/When_deleteCLISecrets_fails_it_should_log_and_continue
  {"level":"info","msg":"Deleting Secrets","namespace":"clusters"}
  {"level":"info","msg":"Failed to delete CLI generated secrets, skipping","error":"failed to clean up secrets in clusters namespace: secrets is forbidden:
  access denied","namespace":"clusters"}
  {"level":"info","msg":"Successfully destroyed cluster and infrastructure","namespace":"clusters","name":"test-cluster","infraID":"test-infra"}
  --- PASS: TestDestroyCluster (0.05s)
      --- PASS: TestDestroyCluster/When_HostedCluster_is_nil_and_platform_specifics_provided_it_should_call_destroyPlatformSpecifics (0.05s)
      --- PASS: TestDestroyCluster/When_kubeconfig_is_set_it_should_use_it_for_the_client (0.00s)
      --- PASS: TestDestroyCluster/When_destroying_a_hosted_cluster_with_platform_specifics,_it_should_set_the_destroy_finalizer_and_complete_successfully
  (0.00s)
      --- PASS: TestDestroyCluster/When_deleteCLISecrets_fails_it_should_log_and_continue (0.00s)
  PASS
  ```

  The `When_deleteCLISecrets_fails_it_should_log_and_continue` subtest injects a fake client
  that returns `403 Forbidden` on `DeleteAllOf` for Secrets, confirming the destroy flow
  logs the error and continues rather than returning it.

  </details>

**Comment by rutvik23** [comment]
After rebasing onto latest changes ^,  re-verified the `hcp destroy cluster` flow on a real cluster built out of HO,CPO images and product-cli from the fix branch, and replicating same kind of RBAC scenario as described in earlier [test report ](https://rutvik23.github.io/my-test-reports/reports/ocpbugs-86669.html) 
<details>

`{"level":"info","ts":"2026-08-12T16:33:14+05:30","msg":"Deleting Secrets","namespace":"clusters"}
{"level":"info","ts":"2026-08-12T16:33:14+05:30","msg":"Failed to delete CLI generated secrets, skipping","error":"failed to clean up secrets in clusters namespace: secrets is forbidden: User \"system:serviceaccount:clusters:hcp-destroy-no-deletecollection\" cannot deletecollection resource \"secrets\" in API group \"\" in the namespace \"clusters\"","namespace":"clusters"}
{"level":"info","ts":"2026-08-12T16:33:14+05:30","msg":"Finalized hosted cluster","namespace":"clusters","name":"finalizer"}
{"level":"info","ts":"2026-08-12T16:33:14+05:30","msg":"Successfully destroyed cluster and infrastructure","namespace":"clusters","name":"finalizer","infraID":"finalizer-jlzvg"}`

**Review by bryan-cox — COMMENTED** [review]
/approve

**Comment by rutvik23** [comment]
/verified by @rutvik23 ([test report reference](https://github.com/openshift/hypershift/pull/8787#issuecomment-5265957148))
